agera/src/main/java/com/google/android/agera/Functions.java

```diff
@@ -110,5 +134,50 @@ public T apply(@NonNull F from) {
     }
   }
 
+  private static final class ParallelMapFunction<F, T>
+      implements Function<List<F>, List<Result<T>>> {
+    @NonNull
+    private final Function<F, T> function;
+    @NonNull
+    private final ExecutorService executorService;
+
+    ParallelMapFunction(@NonNull final ExecutorService executorService,
+        @NonNull final Function<F, T> function) {
+      this.executorService = checkNotNull(executorService);
+      this.function = checkNotNull(function);
+    }
+
+    @NonNull
+    @Override
+    @SuppressWarnings("unchecked")
+    public List<Result<T>> apply(@NonNull final List<F> input) {
+      if (input.isEmpty()) {
+        return emptyList();
+      }
+      final List<Result<T>> result = new ArrayList(input.size());
+      final List<Future<T>> futureResult = new ArrayList(input.size());
+      for (final F item : input) {
+        futureResult.add(executorService.submit(new Callable<T>() {
+          @Override
+          public T call() throws Exception {
+            return function.apply(item);
+          }
+        }));
+      }
+      final int size = futureResult.size();
+      for (int i = 0; i < size; i++) {
+        final Future<T> item = futureResult.get(i);
+        try {
+          result.add(present(item.get()));
+        } catch (final ExecutionException e) {
+          result.add(Result.<T>failure(e.getCause()));
```

Comment:
I'm simply assuming that we'd rather want cause here, and not `ExecutionException`. It's also why I wanted to allow for null in failure. `getCause()` should never return null here but technically it's a nullable method.